File: src/utils/ruleExpander.js
import { DateTime } from 'luxon';

/**
 * Expand schedule rules into concrete per-day intervals for a date range.
 *
 * Each rule is stored in the friend's IANA timezone. We iterate calendar days
 * in that zone, check if the rule fires, then convert the resulting interval
 * to the viewer's zone so the UI can render it as local minutes-since-midnight.
 *
 * @param {Array}  rules       - schedule rule objects
 * @param {string} friendZone  - IANA zone the rule times are expressed in (e.g. 'Europe/London')
 * @param {string} viewerZone  - IANA zone of the person viewing the calendar
 * @param {string} rangeStart  - ISO date 'YYYY-MM-DD' (inclusive, in viewer zone)
 * @param {string} rangeEnd    - ISO date 'YYYY-MM-DD' (inclusive, in viewer zone)
 * @returns {Array} concrete interval objects compatible with the calendar views
 */
export function expandRules(rules, friendZone, viewerZone, rangeStart, rangeEnd) {
  const intervals = [];

  // We iterate days in the FRIEND's zone so weekday checks are correct for them.
  // But we also need to cover days that might spill into the viewer's visible range
  // due to offset differences, so we pad by 1 day on each side.
  let cursor = DateTime.fromISO(rangeStart, { zone: viewerZone }).setZone(friendZone).startOf('day').minus({ days: 1 });
  const end = DateTime.fromISO(rangeEnd, { zone: viewerZone }).setZone(friendZone).endOf('day').plus({ days: 1 });

  while (cursor <= end) {
    const friendDate = cursor.toISODate();          // YYYY-MM-DD in friend's zone
    const friendWeekday = cursor.weekday % 7;       // Luxon weekday: 1=Mon..7=Sun → convert to 0=Sun convention

    for (const rule of rules) {
      if (!matchesDay(rule, friendDate, friendWeekday)) continue;

      const interval = buildInterval(rule, cursor, friendZone, viewerZone);
      if (!interval) continue;

      // Only include if the interval's viewer-side date falls within the requested range
      if (interval.date >= rangeStart && interval.date <= rangeEnd) {
        intervals.push(interval);
      }
    }

    cursor = cursor.plus({ days: 1 });
  }

  return intervals.sort((a, b) =>
    a.date < b.date ? -1 : a.date > b.date ? 1 :
    (a.allDay ? -1 : 0) - (b.allDay ? -1 : 0) ||
    (a.startMin || 0) - (b.startMin || 0)
  );
}

function matchesDay(rule, isoDate, weekdaySun) {
  switch (rule.recurrence) {
    case 'once':
      return rule.date === isoDate;
    case 'daily':
      return true;
    case 'weekly':
      return rule.weekdays && rule.weekdays.includes(weekdaySun);
    default:
      return false;
  }
}

/**
 * Build a concrete interval object from a rule firing on a specific day.
 */
function buildInterval(rule, friendDayStart, friendZone, viewerZone) {
  const id = rule.id + '_' + friendDayStart.toISODate();

  if (rule.allDay) {
    // All-day events: express as the date in the viewer's zone corresponding
    // to the middle of the friend's day
    const midday = friendDayStart.set({ hour: 12 });
    const viewerMidDay = midday.setZone(viewerZone);
    return {
      id,
      ruleId: rule.id,
      friendId: rule.friendId,
      title: rule.title,
      status: rule.status,
      allDay: true,
      startMin: null,
      endMin: null,
      date: viewerMidDay.toISODate(),
      isoStart: friendDayStart.toISO(),
      isoEnd: friendDayStart.plus({ days: 1 }).toISO(),
    };
  }

  // Timed event: build start/end in friend's zone, convert to viewer's zone
  const [sh, sm] = rule.timeStart.split(':').map(Number);
  const [eh, em] = rule.timeEnd.split(':').map(Number);

  const friendStart = friendDayStart.set({ hour: sh, minute: sm, second: 0, millisecond: 0 });
  const friendEnd   = friendDayStart.set({ hour: eh, minute: em, second: 0, millisecond: 0 });

  const viewerStart = friendStart.setZone(viewerZone);
  const viewerEnd   = friendEnd.setZone(viewerZone);

  const viewerDate = viewerStart.toISODate();
  const startMin = viewerStart.hour * 60 + viewerStart.minute;
  const endMin   = viewerEnd.hour * 60 + viewerEnd.minute;

  return {
    id,
    ruleId: rule.id,
    friendId: rule.friendId,
    title: rule.title,
    status: rule.status,
    allDay: false,
    startMin,
    endMin: endMin <= startMin ? 1439 : endMin,   // cap at 23:59 if crosses midnight
    date: viewerDate,
    isoStart: viewerStart.toISO(),
    isoEnd: viewerEnd.toISO(),
  };
}

/**
 * Get the viewer's IANA timezone from the browser.
 */
export function getViewerZone() {
  return Intl.DateTimeFormat().resolvedOptions().timeZone;
}

File: friendule_flutter/lib/utils/rule_expander.dart
import 'package:timezone/data/latest.dart' as tz_data;
import 'package:timezone/timezone.dart' as tz;
import '../models/interval.dart';
import '../models/rule.dart';
import 'date_utils.dart';

bool _tzInitialized = false;

void ensureTimezoneInitialized() {
  if (!_tzInitialized) {
    tz_data.initializeTimeZones();
    _tzInitialized = true;
  }
}

tz.Location getLocation(String? zoneName) {
  ensureTimezoneInitialized();
  if (zoneName == null || zoneName.isEmpty) {
    return tz.getLocation('UTC');
  }
  try {
    return tz.getLocation(zoneName);
  } catch (_) {
    // Fallback if unrecognized
    return tz.getLocation('UTC');
  }
}

bool matchesDay(Rule rule, String isoDate, int weekdaySun) {
  // Check optional date bounds for recurring rules
  if (rule.dateFrom != null && rule.dateFrom!.isNotEmpty && isoDate.compareTo(rule.dateFrom!) < 0) {
    return false;
  }
  if (rule.dateTo != null && rule.dateTo!.isNotEmpty && isoDate.compareTo(rule.dateTo!) > 0) {
    return false;
  }

  switch (rule.recurrence) {
    case 'once':
      return rule.date == isoDate;
    case 'daily':
      return true;
    case 'weekly':
      return rule.weekdays != null && rule.weekdays!.contains(weekdaySun);
    default:
      return false;
  }
}

ScheduleInterval? buildInterval(
  Rule rule,
  tz.TZDateTime friendDayStart,
  tz.Location friendLocation,
  tz.Location viewerLocation,
) {
  final friendIsoDate = ymd(friendDayStart);
  final id = '${rule.id}_$friendIsoDate';

  if (rule.allDay) {
    // All-day events: express as the date in the viewer's zone corresponding
    // to the middle of the friend's day
    final midday = tz.TZDateTime(
      friendLocation,
      friendDayStart.year,
      friendDayStart.month,
      friendDayStart.day,
      12,
      0,
    );
    final viewerMidday = tz.TZDateTime.from(midday, viewerLocation);
    final viewerDate = ymd(viewerMidday);

    return ScheduleInterval(
      id: id,
      ruleId: rule.id,
      friendId: rule.friendId,
      title: rule.title,
      status: rule.status,
      allDay: true,
      startMin: null,
      endMin: null,
      date: viewerDate,
      isoStart: friendDayStart.toIso8601String(),
      isoEnd: friendDayStart.add(const Duration(days: 1)).toIso8601String(),
    );
  }

  // Timed event: build start/end in friend's zone, convert to viewer's zone
  final timeStartStr = rule.timeStart ?? '09:00';
  final timeEndStr = rule.timeEnd ?? '10:00';

  final startParts = timeStartStr.split(':').map(int.parse).toList();
  final endParts = timeEndStr.split(':').map(int.parse).toList();

  final friendStart = tz.TZDateTime(
    friendLocation,
    friendDayStart.year,
    friendDayStart.month,
    friendDayStart.day,
    startParts[0],
    startParts.length > 1 ? startParts[1] : 0,
  );

  final friendEnd = tz.TZDateTime(
    friendLocation,
    friendDayStart.year,
    friendDayStart.month,
    friendDayStart.day,
    endParts[0],
    endParts.length > 1 ? endParts[1] : 0,
  );

  final viewerStart = tz.TZDateTime.from(friendStart, viewerLocation);
  final viewerEnd = tz.TZDateTime.from(friendEnd, viewerLocation);

  final viewerDate = ymd(viewerStart);
  final startMin = viewerStart.hour * 60 + viewerStart.minute;
  var endMin = viewerEnd.hour * 60 + viewerEnd.minute;

  if (endMin <= startMin) {
    // Crosses midnight
    endMin = 1439;
  }

  return ScheduleInterval(
    id: id,
    ruleId: rule.id,
    friendId: rule.friendId,
    title: rule.title,
    status: rule.status,
    allDay: false,
    startMin: startMin,
    endMin: endMin,
    date: viewerDate,
    isoStart: viewerStart.toIso8601String(),
    isoEnd: viewerEnd.toIso8601String(),
  );
}

/// Expand schedule rules into concrete per-day intervals for a date range.
List<ScheduleInterval> expandRules({
  required List<Rule> rules,
  required String friendZone,
  required String viewerZone,
  required String rangeStart,
  required String rangeEnd,
}) {
  ensureTimezoneInitialized();
  final friendLocation = getLocation(friendZone);
  final viewerLocation = getLocation(viewerZone);

  final intervals = <ScheduleInterval>[];

  // We iterate days in the FRIEND's zone so weekday checks are correct for them.
  // Pad by 1 day on each side to cover offset spillover.
  final startDt = parseYmd(rangeStart);
  final endDt = parseYmd(rangeEnd);

  var cursor = tz.TZDateTime(
    friendLocation,
    startDt.year,
    startDt.month,
    startDt.day,
  ).subtract(const Duration(days: 1));

  final end = tz.TZDateTime(
    friendLocation,
    endDt.year,
    endDt.month,
    endDt.day,
    23,
    59,
    59,
  ).add(const Duration(days: 1));

  while (cursor.isBefore(end) || cursor.isAtSameMomentAs(end)) {
    final friendDate = ymd(cursor);
    final friendWeekday = cursor.weekday % 7; // Luxon/Dart 1=Mon..7=Sun -> 0=Sun..6=Sat

    for (final rule in rules) {
      if (!matchesDay(rule, friendDate, friendWeekday)) continue;

      final interval = buildInterval(rule, cursor, friendLocation, viewerLocation);
      if (interval == null) continue;

      // Only include if interval date falls within requested range
      if (interval.date.compareTo(rangeStart) >= 0 && interval.date.compareTo(rangeEnd) <= 0) {
        intervals.add(interval);
      }
    }

    cursor = cursor.add(const Duration(days: 1));
  }

  intervals.sort((a, b) {
    final dateCmp = a.date.compareTo(b.date);
    if (dateCmp != 0) return dateCmp;

    final aAllDay = a.allDay ? -1 : 0;
    final bAllDay = b.allDay ? -1 : 0;
    final allDayCmp = aAllDay.compareTo(bAllDay);
    if (allDayCmp != 0) return allDayCmp;

    final aStart = a.startMin ?? 0;
    final bStart = b.startMin ?? 0;
    return aStart.compareTo(bStart);
  });

  return intervals;
}
